src/nemos/type_casting.py

```diff
@@ -0,0 +1,331 @@
+from numpy.typing import NDArray
+
+from . import utils
+
+
+def is_numpy_array_like(obj) -> bool:
+    """
+    Check if an object is array-like.
+
+    This function determines if an object has array-like properties but isn't an _AstractTsd.
+    An object is considered array-like if it has attributes typically associated with arrays
+    (such as `.shape`, `.dtype`, and `.ndim`), supports indexing, and is iterable.
+
+    Parameters
+    ----------
+    obj : object
+        The object to check for array-like properties.
+
+    Returns
+    -------
+    :
+        True if the object is array-like, False otherwise.
+
+    Notes
+    -----
+    This function uses a combination of checks for attributes (`shape`, `dtype`, `ndim`),
+    indexability, and iterability to determine if the given object behaves like an array.
+    It is designed to be flexible and work with various types of array-like objects, including
+    but not limited to NumPy arrays and JAX arrays. However, it may not be foolproof for all
+    possible array-like types or objects that mimic these properties without being suitable for
+    numerical operations.
+
+    """
+    # Check for array-like attributes
+    has_shape = hasattr(obj, "shape")
+    has_dtype = hasattr(obj, "dtype")
+    has_ndim = hasattr(obj, "ndim")
+
+    # Check for indexability (try to access the first element)
+    try:
+        obj[0]
+        is_indexable = True
+    except (TypeError, IndexError):
+        is_indexable = False
+
+    # Check for iterable property
+    try:
+        iter(obj)
+        is_iterable = True
+    except TypeError:
+        is_iterable = False
+
+    return has_shape and has_dtype and has_ndim and is_indexable and is_iterable
+
+
+def is_pynapple_tsd(x: Any) -> bool:
+    """
+    Verify if an object is a pynapple time series with data object.
+
+    Examines the presence of specific attributes that are characteristic of pynapple time series data structures.
+
+    Parameters
+    ----------
+    x
+        Object to evaluate.
+
+    Returns
+    -------
+    :
+        Indicates the result of the evaluation.
+    """
+    return all(hasattr(x, attr) for attr in ["times", "data"])
+
+
+def _has_same_time_axis(*args, **kwargs) -> bool:
+    """
+    Check for matching time axes among pynapple objects.
+
+    Evaluates whether the time axes of provided pynapple objects are close enough to be considered identical,
+    using a predefined precision level.
+
+    Parameters
+    ----------
+    args:
+        Any positional argument.
+    kwargs:
+        Any keyword argument.
+
+
+    Returns
+    -------
+    :
+    Indicates whether all time axes match.
+    """
+    flat_tree, _ = jax.tree_util.tree_flatten((args, kwargs))
+    if len(flat_tree) == 1:
+        return True
+
+    # get first pynapple
+    is_nap = list(is_pynapple_tsd(x) for x in flat_tree)
+    first_nap = is_nap.index(True)
+
+    # check time samples are close (10**-9 is hard-coded in pynapple)
```

Comment:
or, since it sounds like pynapple doesn't allow you to do "horizontal concatenation" to combine two indepedent `Tsd` objects, and thus doesn't need to check whether time series are identical, maybe we just use the relevant float precision?

Reply:
Pynapple doesn't have a function to check the equivalence of time series. But I can add a parameter in the config file for float precision.

Reply:
Ok you can get it with `nap.config.nap_config.time_index_precision` and it's immutable

Reply:
Until realese I am installing pynapple form the main in github, @gviejo @billbrod, we should remember that at the next pynapple release we should revert that